### PR TITLE
deprecate hydrogen bond analysis in water bridge analysis

### PR DESCRIPTION
**Pull request by xiki-tempula**
Fixes #

deprecate hydrogen bond analysis in water bridge analysis.
The hydrogen bond analysis is not used in the original water bridge analysis and I have cleared some reference to the hydrogen bond analysis in the doc sections.
Move water bridge analysis from hbonds to hydorgenbonds

Related to #2739 and  #2746

Changes made in this Pull Request:
 - 

PR Checklist
------------
 - [ ] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by IAlibay** [comment]
Going to ping @RMeli here, I believe this move was started in PR #2746, but I don't know why it stalled.
(also @xiki-tempula can you add a reference to #2739 in the PR text + eventually changelog? that way we can remember to update the progress on it).

**Comment by xiki-tempula** [comment]
@IAlibay I have chatted with @RMeli before initiating this PR. This PR is still WIP and I will do this after I solved all the PEP8 issue.

**Comment by RMeli** [comment]
@IAlibay, PR #2746 stalled because of https://github.com/MDAnalysis/mdanalysis/issues/2745#issuecomment-644425372 and for lack of time. @xiki-tempula did ask me if he could supersede my PR as part of his current work and I have no problems with that.

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/hbonds/wbridge_analysis.py:44]: So in the past we've been a little bit more cautious with stubs, specifically setting simplewarning through something like:

```python
with warnings.catch_warnings():
    warnings.simplefilter('always', DeprecationWarning)
    warnings.warn(("This module has been moved to "
                   "MDAnalysis.analysis.hydrogenbonds.wbridge_analysis "
                   "It will be removed in release 3.0"),
                  category=DeprecationWarning)

from ..hydrogenbonds.wbridge_analysis import WaterBridgeAnalysis
```

Whilst I'm not sure the order of the import matters so much, the use of filtering is probably useful in cases where people tend to just blanket remove all warnings. Pinging @orbeckst here who probably has more context on why this was the decision taken for stubs circa ~ v0.21.

File: package/MDAnalysis/analysis/hydrogenbonds/__init__.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/hydrogenbonds/__init__.py:26]: Also add to `__all__`?
```suggestion
    'WaterBridgeAnalysis'
]
```

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:13]: Also needs a test for the new warning please.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1]: From a cursory look, it seems fine as just the moved file + PEP8 changes. Unfortunately the relatively unhelpful git diff make this hard to check thoroughly (PEP8 probably could have just been dealt with in a separate PR but too late now).

Is there any particular changes here that you'd like us to look at?

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:698]: Unnecessary white space?

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/hbonds/wbridge_analysis.py:46]: Next target version is 2.0, so removals have to be done afterwards unless we want to make this a v1.1 target instead. Thoughts @orbeckst ?

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple of comments + need to work out what the target version will be for full removal.

File: package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py
Comment (xiki-tempula) [reply package/MDAnalysis/analysis/hydrogenbonds/wbridge_analysis.py:1]: As is described in the PR description. I have 

- Fixed the PEP8 issue
- Remove reference to the old HBA in the doc
- Added the code for deprecation related stuff.

None of the code inside the wba is modified.

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:24]: This is going to go away next, so I probably wouldn't test for it.

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:20]: Given my other comment, rather than explicitly testing the record, it's probably easier/cleaner to just check using the `match` keyword here:
```suggestion
    wmsg = 'Please use MDAnalysis.analysis.hydrogenbonds.wbridge_analysis'

    with pytest.warns(DeprecationWarning, match=wmsg):
```

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:38]: Strange, does it not work if this import is at the top of the file?

Comment (xiki-tempula) [reply testsuite/MDAnalysisTests/analysis/test_wbridge.py:38]: @IAlibay  The warning is only issued once when it is first imported. Importing it again won't give the same warning.

Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_wbridge.py:27]: I see, well as long as it works, it's a lot better than whatever I had to do for some of the other tests with import issues 😅

Comment (xiki-tempula) [reply testsuite/MDAnalysisTests/analysis/test_wbridge.py:27]: @IAlibay but it creates PEP8 problem as the module-level import is not at the top of the file.

**Comment by IAlibay** [comment]
Thanks for all the work here @xiki-tempula,  just waiting on a secondary opinion on some of the comments I left above. Particularly, I'm not 100% sure if we should not be targeting 1.x instead of 2.0.

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_wbridge.py:27]: Unfortunately I can't pull your PR (since it's on your develop branch), so I can't tell if there is some weird way to go around this, but if this is the only way I think we can live with breaking PEP8 a little bit (it's more guidelines rather than strict rules to live by).

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py
Comment (lilyminium) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py:46]: Are we doing more versions of 1? I would deprecate in 1.xx and remove in 2.0, as was done for #2791 and #2622.

Comment (orbeckst) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py:46]: If we have to do a 1.1 for whatever reasons then we keep things as unchanging as possible. But the goal was to keep 1.x boring and as similar to the old stuff as possible, so any removals target 2.0.

I _hope_ that we are only doing bug fixes for 1.x so that this stays 1.0.x

File: package/MDAnalysis/analysis/hbonds/wbridge_analysis.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/hbonds/wbridge_analysis.py:44]: Sorry for the late (and unhelpful) reply: I don't remember and the more cautious approach by @IAlibay looks good to me.

Removal would be 2.0 (together with old h-bonds).

File: package/CHANGELOG
Comment (orbeckst) [inline package/CHANGELOG:126]: indentation
```suggestion
    hydrogen bond analysis in water bridge analysis (Issue #2739 PR #2913)
```

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_wbridge.py:3]: Why do we need `reload`? This is very unusual and I'd not rather have it unless absolutely necessary. 

It is not explicitly used so should be removed.

**Review by orbeckst — CHANGES_REQUESTED** [review]
This PR should really be against master to be part of release 1.0.1 #2768 . The deprecations should become part of 1.0.1 and the removal is scheduled for 2.0.

You then need a second PR against develop that does _not_ contain warnings etc and only contains the moved module.

I am not sure how you want to organize this in the best way – maybe make a second PR that targets master and then remove the warnings/deprecations from this one (and change the CHANGELOG accordingly).

File: testsuite/MDAnalysisTests/analysis/test_wbridge.py
Comment (xiki-tempula) [reply testsuite/MDAnalysisTests/analysis/test_wbridge.py:3]: The reload is for the checking of warnings.
The warning will only appear upon the import as I use reload to regenerate the warning so that it could be checked.

**Comment by xiki-tempula** [comment]
@orbeckst Thank you for the advice. I have changed the changelog and remove the water_bridge from the hbond module. I will open another PR for master.

**Comment by xiki-tempula** [comment]
@orbeckst Thank you for the advice. I was wondering if there is anything that I could do to get this PR merged. I'm working to port @p-j-smith 's MDAnalysis.analysis.hydrogenbonds.hbond_analysis.HydrogenBondAnalysis.lifetime() to water bridge analysis, so it would be nice if this PR could be merged.

**Comment by orbeckst** [comment]
Sorry , I am snowed under with other things at the moment. Could you please ping me at the end of the week again? Sorry again for the long silence.

**Comment by xiki-tempula** [comment]
@orbeckst Thanks for all the help. I wonder if you mind give a look, please? I will make another PR to the masters once this one is merged.

**Comment by orbeckst** [comment]
Thanks for the reminder. I put it in my calendar for Wednesday — sorry, there’s no way I can get to it any earlier (and I had not enough spare time last week).

**Review by orbeckst — CHANGES_REQUESTED** [review]
This PR moves wbridge_analysis to hydrogenbonds. Please open a PR against master that just includes the deprecations, stating that in 2.0, it will be in hydrogenbonds. When the deprecation PR is approved, I can also approve the move here.

**Comment by orbeckst** [comment]
@xiki-tempula what is your plan for wbridge analysis? Do you want to rewrite it to base it on the new hydrogenbond analysis? It would be really good to have HydrogenBondAnalysis and WaterBridgeAnalysis behave in the same way as far as selections and performance go. 

If there's a way to avoid duplication of code then I would be very much in favor of, e.g., creating a base class for both HydrogenBondAnalysis and WaterBridgeAnalysis.

**Comment by xiki-tempula** [comment]
@orbeckst The PR against the master is done in #3111.

The issue with HydrogenBondAnalysis is a quite tricky one. Since the WaterBridgeAnalysis is written based on the old HydrogenBondAnalysis, it has all the functionality of the old HydrogenBondAnalysis. 

Later on, @p-j-smith introduced capped-distance to HydrogenBondAnalysis, which brought a major speed improvement. So I have adopted the capped-distance to distance calculations. Currently, WaterBridgeAnalysis and HydrogenBondAnalysis offer the same performance in terms of computing hydrogen bonds.

The selection problem is a bit tricky. As is discussed in #2177, the WaterBridgeAnalysis adopted the same selection scheme as the old HydrogenBondAnalysis. I'm currently working on a RDKIT based selection scheme.

It is difficult to write WaterBridgeAnalysis based on HydrogenBondAnalysis as WaterBridgeAnalysis needs a more complex data structure but it is easy to write HydrogenBondAnalysis based on WaterBridgeAnalysis. The only functionality that is present in HydrogenBondAnalysis but not WaterBridgeAnalysis is the lifetime analysis, which I plan to add when this PR is merged.

**Comment by p-j-smith** [comment]
Hi, I don't know the full extent of the functionality of the WaterBridgeAnalysis class, but I have used HydrogenBondAnalysis to find hydrogen bonds and then used NetworkX to look for water-bridging in the following way:

1. Find hydrogen bonds. Use an selection that include e.g. protein and water within 10 Angstrom of the protein, such as in the fourth example seen here: https://docs.mdanalysis.org/2.0.0-dev0/documentation_pages/analysis/hydrogenbonds.html#example-use-of-hydrogenbondanalysis Using the 10 Angstrom cutoff (as opposed to all water) makes both the hydrogen bond analysis and the path-finding faster.

2. Iterate over frames. At each frame construct an adjacency matrix, A, of size (num_donors, num_accepors), where Aij=1 if there is a hydrogen bond between donor_i and acceptor_j. Use NetworkX to construct a graph from this adjacency matrix. Iterate over donor-acceptor pairs of the protein only, and use the NetworkX `all_shorted_paths` method (https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.generic.all_shortest_paths.html#networkx.algorithms.shortest_paths.generic.all_shortest_paths) to find the shortest water-mediated paths between each donor and acceptor, if one exists. If the path is greater than a cutoff, say 5 water molecules, don't count this as path. I'm not sure, but I think there is also an all-to-all method in NetworkX that could remove the need to iterate over donor-acceptor pairs.

Another PhD student in my group had the idea of using NetworkX for this. We've only used it to look at the statistics of the path-length distribution, but NetworkX also returns the list of atoms in the through-water path from donor-to acceptor. I'm not sure whether this is something you would be interested in, but offloading the path-finding to NetworkX might give a performance boost whilst also keeping the same selections etc. as HydrogenBondAnalysis. If this is something you're interested in @xiki-tempula I'd be happy to work with you on it.

**Comment by xiki-tempula** [comment]
@p-j-smith Thanks for the insight. The WaterBridgeAnalysis basically does what you have said with various tweaks for speed improvement and robustness.

The first step is to find the hydrogen bonds within a defined space. Instead of `protein and water within 10 Angstrom of the protein`, I make three selections, one selection is the start selection (selection 1) another selection is the end selection (selection 2), while the third water selection is a convex which is both within a certain distance from selection 1 and selection 2. Since the hydrogen bond from selection 1 to selection 2 is implicitly detected in this stage, it is easy to go from WaterBridgeAnalysis to HydrogenBondAnalysis but not the other way around.

The second step is finding a path from selection 1 to selection 2. I have written an optimized path-finder for this type of job. Though I have not tested the NetworkX implementation, I think it would be interesting to see how would NetworkX perform.
I don't think NetworkX would be significantly slower than my implementation but most of the time is spent on distance search, so it is not likely to matter a lot.

I think one strength of the WaterBridgeAnalysis is the flexible analysis functionality. The `path-length distribution` could be easily computed by providing the custom analysis function to the count_by_type function.

**Comment by IAlibay** [comment]
> The selection problem is a bit tricky. As is discussed in #2177, the WaterBridgeAnalysis adopted the same selection scheme as the old HydrogenBondAnalysis. I'm currently working on a RDKIT based selection scheme.

Just jumping in here and focusing very specifically on the last this specifically. I don't know what your plans are re: RDKIT-based selections, but if it's possible to make it optional that would be great. WaterBridgeAnalysis is a great core analysis feature of MDA, since we probably can't make RDKIT a core dependency, having the ability to use it without RDKIT is probably quite important.

**Comment by p-j-smith** [comment]
> I make three selections, one selection is the start selection (selection 1) another selection is the end selection (selection 2), while the third water selection is a convex which is both within a certain distance from selection 1 and selection 2.

Sorry - it's been a while since I worked on this. You're completely right - I remember now that in post-processing the hydrogen bonds to find water-bridging I also had to use three selections - for the source, sink and path. So what I suggested would probably complicate rather than simplify things.

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
With PR #3111 to be merged into 1.0.2-develop, this is good to go from my end.

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG:159]: Technically new changes go to the top of the list, but I'm willing to let that go.

**Review by IAlibay — APPROVED** [review]
Sorry, I've not been keeping up, from a cursory look it seems fine. I'll approve since @orbeckst did / so I'm not blocking merging :)

**Comment by orbeckst** [comment]
I want to rebase all of this into a single commit (there's too much clutter in the history) but *Rebase and merge* will not work automatically.

@xiki-tempula please condense the history of this PR in not more than ~3 commits (or just a single one) and then force-push so that we have a clean history. I leave this to you because you know best what the end result ought to look like.

I will block until the history is cleaned up.

**Comment by orbeckst** [comment]
Actually, forget my previous comment: *Squash and merge* works just fine. Will do this *now*. Sorry for the clutter!

File: package/CHANGELOG
Comment (orbeckst) [reply package/CHANGELOG:159]: I appreciate your attention to detail.

I merged the whole thing nevertheless, as this has been sitting for way too long.

**Pull request merged**
